File: src/optimizers/nadam.jl
@defstruct NadamOptions <: AbstractOptimizerOptions (
  (lr                :: Real = 0.001, lr > 0),
  (beta1             :: Real = 0.99,  beta1 > 0 && beta1 < 1),
  (beta2             :: Real = 0.999,  beta2 > 0 && beta2 < 1),
  (epsilon           :: Real = 1e-8, epsilon > 0),
  (grad_clip         :: Real = 0, grad_clip >= 0),
  (weight_decay      :: Real = 0.00001, weight_decay >= 0),
  lr_scheduler       :: Any = nothing,
  momentum_scheduler :: Any = nothing
)

"""
    Nadam

Nesterov Adam optimizer: Adam RMSprop with Nesterov momentum,
see [1] and notes for further description.

    Nadam(; kwargs...)

# Attributes
* `lr::Real`: default `0.001`, learning rate.
* `beta1::Real`: default `0.99`.
* `beta2::Real`: default `0.999`.
* `epsilon::Real`: default `1e-8`, small value added for
  numerical stability
* `grad_clip::Real`: default `0`, if positive, will clip the gradient
  into the range `[-grad_clip, grad_clip]`.
* `weight_decay::Real`: default `0.00001`, weight decay is equivalent
  to adding a global l2 regularizer for all the parameters.
* `lr_scheduler::AbstractLearningRateScheduler`: default `nothing`, a
  dynamic learning rate scheduler. If set, will overwrite the `lr`
  parameter.
* `momentum_scheduler::AbstractMomentumScheduler` default
  `NadamScheduler` of the form
  ``\mu_t = beta1 * (1 - 0.5 * 0.96^{t * 0.004})``

# Notes
Default parameters follow those provided in the paper.
It is recommended to leave the parameters of this optimizer
at their default values.

# References
* [1]: Incorporating Nesterov Momentum into Adam.
  [http://cs229.stanford.edu/proj2015/054_report.pdf]
  (http://cs229.stanford.edu/proj2015/054_report.pdf)
* [2]: On the importance of initialization and momentum in deep learning
  [http://www.cs.toronto.edu/~fritz/absps/momentum.pdf]
  (http://www.cs.toronto.edu/~fritz/absps/momentum.pdf)
"""
type Nadam <: AbstractOptimizer
  opts  :: NadamOptions
  state :: OptimizationState

  function Nadam(; kwargs...)
    opts = NadamOptions(; kwargs...)
    opts.lr_scheduler = get_lr_scheduler(opts.lr_scheduler, opts.lr)
    opts.momentum_scheduler = get_momentum_scheduler(opts.momentum_scheduler,
      Momentum.NadamScheduler(mu0=opts.beta1))

    new(opts)
  end
end

type NadamState
  mt         :: NDArray
  nt         :: NDArray
  momentum   :: Float64
  beta2Power :: Float64
end

function create_state(self :: Nadam, index :: Int, weight :: NDArray)
  return NadamState( zeros(size(weight), context(weight)),
                     zeros(size(weight), context(weight)),
                     1.0,
                     self.opts.beta2 )
end

function update(self :: Nadam, index :: Int, weight :: NDArray,
                grad :: NDArray, state :: NadamState)
  lr = get_learning_rate(self.opts.lr_scheduler, self.state)
  grad = normalized_gradient(self.opts, self.state, weight, grad)

  mu_t, mu_t1 =
    get_momentum(self.opts.momentum_scheduler, self.state)
  state.momentum *= mu_t
  momentum_next = state.momentum * mu_t1

  grad_prime = grad / (1.0 - state.momentum)
  @inplace state.mt .*= self.opts.beta1
  @inplace state.mt .+= (1.0 - self.opts.beta1) * grad
  mt = state.mt / (1.0 - momentum_next)

  @inplace state.nt .*= self.opts.beta2
  @inplace state.nt .+= (1.0 - self.opts.beta2) * grad .* grad
  nt = state.nt / (1.0 - state.beta2Power)
  state.beta2Power *= self.opts.beta2

  mt_prime = (1.0 - mu_t) * grad_prime + mu_t1 * mt
  @inplace weight .+= -lr * mt_prime ./ (sqrt(nt) + self.opts.epsilon)
end

File: python/mxnet/optimizer/nadam.py
# coding: utf-8
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# pylint: disable=W0223
"""Nadam optimizer."""
from __future__ import absolute_import
from ..ndarray import (zeros, clip, sqrt, square)
from .optimizer import Optimizer, register

__all__ = ['Nadam']


@register
class Nadam(Optimizer):
    """The Nesterov Adam optimizer.

    Much like Adam is essentially RMSprop with momentum,
    Nadam is Adam RMSprop with Nesterov momentum available
    at http://cs229.stanford.edu/proj2015/054_report.pdf.

    This optimizer accepts the following parameters in addition to those accepted
    by :class:`.Optimizer`.

    Parameters
    ----------
    learning_rate : float, default 0.001
        The initial learning rate. If None, the optimization will use the
        learning rate from ``lr_scheduler``. If not None, it will overwrite
        the learning rate in ``lr_scheduler``. If None and ``lr_scheduler``
        is also None, then it will be set to 0.01 by default.
    beta1 : float, default 0.9
        Exponential decay rate for the first moment estimates.
    beta2 : float, default 0.999
        Exponential decay rate for the second moment estimates.
    epsilon : float, default 1e-8
        Small value to avoid division by 0.
    schedule_decay : float, default 0.004
        Exponential decay rate for the momentum schedule
    use_fused_step : bool, default False
        Whether or not to use fused kernels for optimizer.
        When use_fused_step=False, step is called,
        otherwise, fused_step is called.
    """
    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8,
                 schedule_decay=0.004, use_fused_step=False, **kwargs):
        super(Nadam, self).__init__(learning_rate=learning_rate,
                                    use_fused_step=use_fused_step,
                                    **kwargs)
        self.beta1 = beta1
        self.beta2 = beta2
        self.epsilon = epsilon
        self.schedule_decay = schedule_decay
        self.m_schedule = 1.

    def create_state(self, index, weight):
        return (zeros(weight.shape, weight.context, dtype=weight.dtype),  # mean
                zeros(weight.shape, weight.context, dtype=weight.dtype))  # variance

    def step(self, indices, weights, grads, states):
        """Perform an optimization step using gradients and states.

        Parameters
        ----------
        indices : list of int
            List of unique indices of the parameters into the individual learning rates
            and weight decays. Learning rates and weight decay may be set via `set_lr_mult()`
            and `set_wd_mult()`, respectively.
        weights : list of NDArray
            List of parameters to be updated.
        grads : list of NDArray
            List of gradients of the objective with respect to this parameter.
        states : List of any obj
            List of state returned by `create_state()`.
        """
        for index, weight, grad, state in zip(indices, weights, grads, states):
            self._update_count(index)
            lr = self._get_lr(index)
            wd = self._get_wd(index)
            t = self._index_update_count[index]

            # preprocess grad
            grad *= self.rescale_grad
            if self.clip_gradient is not None:
                grad = clip(grad, -self.clip_gradient, self.clip_gradient)
            grad += wd * weight

            coef2 = 1. - self.beta2**t

            # warming momentum schedule
            momentum_t = self.beta1 * (1. - 0.5 * (pow(0.96, t * self.schedule_decay)))
            momentum_t_1 = self.beta1 * (1. - 0.5 * (pow(0.96, (t + 1) * self.schedule_decay)))
            self.m_schedule = self.m_schedule * momentum_t
            m_schedule_next = self.m_schedule * momentum_t_1

            # update mean and var
            mean, var = state
            mean[:] *= self.beta1
            mean[:] += (1. - self.beta1) * grad
            var[:] *= self.beta2
            var[:] += (1. - self.beta2) * square(grad)

            grad_prime = grad / (1. - self.m_schedule)
            mean_prime = mean / (1. - m_schedule_next)
            var_prime = var / coef2
            mean_bar = momentum_t_1 * mean_prime + (1. - momentum_t) * grad_prime

            # update weight
            d = mean_bar / (sqrt(var_prime) + self.epsilon)
            weight[:] -= lr * d
